### PR TITLE
Reenable CSharpIsAndCastCheckWithoutNameDiagnosticAnalyzer

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/22531

This feature was disabled because is triggered poor behavior in NameGenerator (where tons of memory was used due to inefficient linq calls).  That behavior was fixed by me in https://github.com/dotnet/roslyn/pull/39974.  As such, I'm reenabling the feature.

### CODE REVIEW

File: src/Features/CSharp/Portable/UsePatternMatching/CSharpIsAndCastCheckWithoutNameDiagnosticAnalyzer.cs [hunk 1]
```diff
@@ -28,11 +29,7 @@ namespace Microsoft.CodeAnalysis.CSharp.UsePatternMatching
     /// but only for code cases where the user has provided an appropriate variable name in
     /// code that can be used).
     /// </summary>
-    //
-    // disabled for preview 1 due to some perf issue. 
-    // we will re-enable it once the issue is addressed.
-    // https://devdiv.visualstudio.com/DevDiv/_workitems?id=504089&_a=edit&triage=true 
-    // [DiagnosticAnalyzer(LanguageNames.CSharp), Shared]
+    [DiagnosticAnalyzer(LanguageNames.CSharp), Shared]
     internal class CSharpIsAndCastCheckWithoutNameDiagnosticAnalyzer : AbstractBuiltInCodeStyleDiagnosticAnalyzer
```
Comment (mavasani) [inline src/Features/CSharp/Portable/UsePatternMatching/CSharpIsAndCastCheckWithoutNameDiagnosticAnalyzer.cs]: Are the tests already enabled?

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/UsePatternMatching/CSharpIsAndCastCheckWithoutNameDiagnosticAnalyzer.cs]: Yes.

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/UsePatternMatching/CSharpIsAndCastCheckWithoutNameDiagnosticAnalyzer.cs]: https://github.com/dotnet/roslyn/blob/master/src/EditorFeatures/CSharpTest/UsePatternMatching/CSharpIsAndCastCheckWithoutNameDiagnosticAnalyzerTests.cs

Our tests don't validate actual proper exporting.  tehy just call directly into the analyzers/fixers. I opened an issue about that oen time, but i can't find it :-/

Comment (sharwell) [reply src/Features/CSharp/Portable/UsePatternMatching/CSharpIsAndCastCheckWithoutNameDiagnosticAnalyzer.cs]: > Our tests don't validate actual proper exporting

I've seen a real bug for this in the past too.

**Pull request merged**